### ISSUE TITLE
Torch Compile mode giving Graph Break for Higher Order Op for ProcessGroup param

### ISSUE THREAD

**Issue by amitchawla1** [pytorch/pytorch#129942]
### 🐛 Describe the bug

When using compile mode, i am getting graph breaks due with following errors

"[rank0]:I0702 10:23:29.307000 139872906987520 torch/_dynamo/variables/higher_order_ops.py:468] [0/0] autograd.Function with body that accepts non-Tensors as input. Got: <class 'torch.distributed.distributed_c10d.ProcessGroup'>"

### Versions

Collecting environment information...
PyTorch version: 2.3.1a0+gita805a07
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.5 (ssh://gerrit.habana-labs.com:29418/tpc_llvm10 f38648243e255fbc4d3e104d9acb78ef810528da)
CMake version: version 3.27.7
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 1
Core(s) per socket:                 6
Socket(s):                          2
Stepping:                           0
BogoMIPS:                           5187.81
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  VMware
Virtualization type:                full
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           12 MiB (12 instances)
L3 cache:                           38.5 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Syscall hardening, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] habana-torch-dataloader==1.17.0+git1606a40eb
[pip3] habana-torch-plugin==1.17.0+git1606a40eb
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] optree==0.11.0
[pip3] pytorch-lightning==2.3.0
[pip3] torch==2.3.1a0+gita805a07
[pip3] torch_tb_profiler==0.4.0
[pip3] torchaudio==2.3.0+952ea74
[pip3] torchdata==0.7.1+5e6f7b7
[pip3] torchmetrics==1.4.0.post0
[pip3] torchtext==0.18.0a0+9bed85d
[pip3] torchvision==0.18.1a0+fe70bc8
[conda] Could not collect

**Comment by amitchawla1** [comment]
`import torch
import torch.multiprocessing as mp
import torch.distributed as dist
from typing import Callable, Dict, TYPE_CHECKING, Any, Optional, Tuple, Union
from torch import Tensor
import os

def get_comm() -> dist.ProcessGroup:
    return pg

class AllToAll(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        input = _AllToAll.apply(get_comm(), input)
        return 0.5 * (5 * input ** 3 - 3 * input)

    @staticmethod
    def backward(ctx, grad_output):
        input, = ctx.saved_tensors
        return grad_output * 1.5 * (5 * input ** 2 - 1)

class _AllToAll(torch.autograd.Function):
    @staticmethod
    def forward(ctx: Any, group: dist.ProcessGroup, input: Tensor) -> Tensor:  # type: ignore
        ctx.group = group
        input = input.contiguous()
        output = torch.empty_like(input)
        dist.all_to_all_single(output, input, group=group)
        return output

    @staticmethod
    def backward(ctx: Any, *grad_output: Tensor) -> Tuple[None, Tensor]:
        return (None, _AllToAll.apply(ctx.group, *grad_output))

def all_to_all(input):
    input1 = input.relu()
    input1 = AllToAll.apply(input1)
    input1.sum().backward()
    return input1

class DistSetup:
    def __init__(self, rank, world_size):
        global pg
        os.environ["MASTER_ADDR"] = "localhost"
        os.environ["MASTER_PORT"] = "12345"
        dist.init_process_group(backend="gloo", rank=rank, world_size=world_size)
        pg = dist.new_group(backend="gloo")

    def __del__(self):
        dist.destroy_process_group()

def main_func(gpu_rank, world_size):
    _ = DistSetup(gpu_rank, world_size)
    compile_func = torch.compile(all_to_all)
    input = torch.arange(world_size) + gpu_rank * world_size
    input = input.float()
    input.requires_grad = True
    compile_func(input.to('cpu'))

if __name__ == "__main__":
    n_gpus = 4
    mp.spawn(main_func, args=(n_gpus, ), nprocs=n_gpus, join=True)`

**Comment by wbigat2** [comment]
It seems that ProcessGroupVariable cannot be expressed in fx graph?

**Comment by ezyang** [comment]
No, our general stance has been that there is one PG and we don't have to represent it in the graph.

**Comment by wbigat2** [comment]
> No, our general stance has been that there is one PG and we don't have to represent it in the graph.

Oh, I understand that the solution to this kind of problem is that the user should switch to the new collective communication interface in the script, such as c10d_functional.all_to_all_single,  instead of distributed.all_to_all_single  which also uses PG as input.
Am I right?  @ezyang

**Comment by ezyang** [comment]
Yes. Actually, Dynamo should do this translation automatically, so if it's not that's also a bug.

**Comment by wbigat** [comment]
> Yes. Actually, Dynamo should do this translation automatically, so if it's not that's also a bug.

I get it, thank you very much.

**Comment by LucasLLC** [comment]
@wanchaol , @wz337 would you mind taking a look?

**Comment by amitchawla1** [comment]
@wz337, @wanchaol Any plans for this ?

**Comment by anijain2305** [comment]
cc @bdhirsh

**Comment by masnesral** [comment]
I'm helping scrub old issues. Looks like we never responded about plans. @bdhirsh ?

**Comment by bdhirsh** [comment]
I guess the current state here is that:

(1) dynamo has *some* logic for converting non-functional, process-group-based-collectives into functional collectives

(2) But this converter logic is not smart enough to convert an entire `autograd.Function` that accepts a `ProcessGroup` input and saves it for backward.

It probably makes sense to leave this issue open? Although hopefully you can just work around, if you are care about torch.compile support, by directly using the functional collectives instead. See `all_to_all_single` here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L451
